[Build] ScheduleTree struct error schedule_tree.h

- OS: Fedora 28
- Python version: 3.6
- CUDA: 9.1
- GCC: 8.0.1

Hi Folks,

I am trying to pack TC as rpm package on RedHat/Fedora platform but have issues within **tc_executor.cc** & **schedule_tree.h** . Not sure what would be right declartion.

* In short:
```
polyhedral/schedule_tree.h:141:44: error: ‘std::ostream& tc::polyhedral::detail::operator<<(std::ostream&, tc::polyhedral::detail::ScheduleTree&)’ should have been declared inside ‘tc::polyhedral::detail’
```

* In details
```
/usr/bin/c++  -DCAFFE2_USE_GOOGLE_GLOG -DDMLC_USE_GLOG -Dtc_core_EXPORTS -I/home/cbalint/rpmbuild/BUILD/TensorComprehensions/src/../include -I/home/cbalint/rpmbuild/BUILD/TensorComprehensions/src/. -I/usr/lib/python2.7/site-packages/torch/lib/include -I/home/cbalint/rpmbuild/BUILD/TensorComprehensions/third-party/caffe2/third_party/eigen -I/usr/include/cuda -I/usr/include/cuda/include -I/home/cbalint/rpmbuild/BUILD/TensorComprehensions/third-party/dlpack/include -I/home/cbalint/rpmbuild/BUILD/TensorComprehensions/third-party/islpp/include -I/home/cbalint/rpmbuild/BUILD/TensorComprehensions/include -I/home/cbalint/rpmbuild/BUILD/TensorComprehensions/src/version -I/home/cbalint/rpmbuild/BUILD/TensorComprehensions/build/src/proto  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -mcet -fcf-protection -DCUDA_HOME="\"/usr/include/cuda\"" -DCUB_HOME="\"/usr/include\""  -Wall -Wno-sign-compare -fPIC  -o CMakeFiles/tc_core.dir/tc_executor.cc.o -c /home/cbalint/rpmbuild/BUILD/TensorComprehensions/src/core/tc_executor.cc -std=gnu++11
In file included from /home/cbalint/rpmbuild/BUILD/TensorComprehensions/src/../include/tc/core/halide2isl.h:25,
                 from /home/cbalint/rpmbuild/BUILD/TensorComprehensions/src/../include/tc/core/polyhedral/scop.h:27,
                 from /home/cbalint/rpmbuild/BUILD/TensorComprehensions/src/../include/tc/core/tc_executor.h:20,
                 from /home/cbalint/rpmbuild/BUILD/TensorComprehensions/src/core/tc_executor.cc:16:
/home/cbalint/rpmbuild/BUILD/TensorComprehensions/src/../include/tc/core/polyhedral/schedule_tree.h:141:44: error: ‘std::ostream& tc::polyhedral::detail::operator<<(std::ostream&, tc::polyhedral::detail::ScheduleTree&)’ should have been declared inside ‘tc::polyhedral::detail’
       tc::polyhedral::detail::ScheduleTree&);
```

--- a/include/tc/core/polyhedral/schedule_tree.h
+++ b/include/tc/core/polyhedral/schedule_tree.h
@@ -135,6 +135,9 @@ namespace detail {
 // [1] Verdoolaege, Guelton, Grosser & Cohen (2014). "Schedule trees". In
 // IMPACT 2014.
 //////////////////////////////////////////////////////////////////////////////
+
+std::ostream& operator<<(std::ostream& os, const ScheduleTree& tree);
+
 struct ScheduleTree {
   friend std::ostream& tc::polyhedral::detail::operator<<(
       std::ostream&,
@@ -509,8 +512,6 @@ struct ScheduleTree {
   detail::ScheduleTreeType type_{detail::ScheduleTreeType::None};
   std::unique_ptr<ScheduleTreeElemBase> elem_{nullptr};
 };
-
-std::ostream& operator<<(std::ostream& os, const ScheduleTree& tree);
 
 } // namespace detail
 

